renderdoc/driver/shaders/spirv/spirv_debug_setup.cpp

```diff
@@ -1685,109 +1871,278 @@ void Debugger::FillDebugSourceVars(rdcarray<InstructionSourceInfo> &instInfo)
-              for(uint32_t x = 0; x < sourceVar.rows * sourceVar.columns; x++)
-                sourceVar.variables.push_back(DebugVariableReference(
-                    DebugVariableType::Variable, GetRawName(mapping.debugVar), x));
+              RDCASSERTEQUAL(countRemainingIndexes, 0);
+              // Remove mappings : this mapping covers everything
+              usage->debugVar = mapping.debugVar;
+              usage->children.clear();
+              usage->emitSourceVar = true;
+              usage->regSuffix.clear();
             }
           }
           else
           {
-            // walk down until we get to a scalar type, if we get there. This means arrays of basic
-            // types will get the right type
+            // walk down until we get to a scalar type, if we get there. This means arrays of
+            // basic types will get the right type
             while(typeWalk && typeWalk->baseType != Id() && typeWalk->type == VarType::Unknown)
               typeWalk = &m_DebugInfo.types[typeWalk->baseType];
 
-            sourceVar.type = typeWalk->type;
-
-            // anything else we treat as a struct
-            if(sourceVar.type == VarType::Unknown)
-              sourceVar.type = VarType::Struct;
-
-            sourceVar.variables.push_back(DebugVariableReference(DebugVariableType::Variable,
-                                                                 GetRawName(mapping.debugVar), 0));
+            usage->type = typeWalk->type;
+            usage->debugVar = mapping.debugVar;
+            usage->debugVarComponent = 0;
+            usage->rows = 1U;
+            usage->columns = 1U;
+            usage->emitSourceVar = true;
+            usage->children.clear();
+            usage->regSuffix.clear();
           }
-
-          i.sourceVars.push_back(sourceVar);
         }
-
-        processed.push_back(mapping);
       }
+    }
 
-      // if we reach a function scope, don't go up any further.
-      if(scope->type == DebugScope::Function)
-        break;
+    for(size_t sv = 0; sv < sourceVars.size(); ++sv)
+    {
+      Id sourceVarId = sourceVars[sv];
+      Node *usage = &roots[sourceVarId];
+      rdcarray<const Node *> nodesToProcess;
+      rdcarray<const Node *> sourceVarNodes;
+      nodesToProcess.push_back(usage);
+      while(!nodesToProcess.isEmpty())
+      {
+        const Node *n = nodesToProcess.back();
+        nodesToProcess.pop_back();
+        if(n->emitSourceVar)
+        {
+          sourceVarNodes.push_back(n);
+        }
+        else
+        {
+          for(size_t x = 0; x < n->children.size(); ++x)
+          {
+            const Node *child = &n->children[x];
+            nodesToProcess.push_back(child);
+          }
+        }
+      }
+      for(size_t x = 0; x < sourceVarNodes.size(); ++x)
+      {
+        const Node *n = sourceVarNodes[x];
+        SourceVariableMapping sourceVar;
+        sourceVar.name = n->name;
+        sourceVar.offset = 0;
+        if(n->columnMajor)
```

Comment:
This is not tested or verified and matches the existing code.
`DXC` / `glslang` do not generate `DebugTypeMatrix` instructions
`slang` does generate `DebugTypeMatrix` instruction but I am not sure `slang` generation is correct.